### PR TITLE
Add API Reference to website

### PR DESCRIPTION
**Pull request by clemyan**
**What's the problem this PR addresses?**

The API reference is not available in new website

**How did you fix it?**

This is a WIP to bring the API reference to the new website via `docusaurus-plugin-typedoc-api`.

The current state is has some kinks and issues to work out. The purpose of this draft PR is to track and discuss. I'll sort those out over time but if anyone wants to take over then be my guest.

- ~~Seems like `docusaurus-plugin-typedoc-api` need to use the instance from `@docusaurus/theme-common` and `@docusaurus/plugin-content-docs` used by the main doc, which were hidden inside `@docusaurus/preset-classic`. I added package extensions to resolve them but that seems brittle.~~
- ~~Docusaurus's MDX loader resolves `react` and `@mdx-js/react` from the loaded Markdown file. To include READMEs into the API reference we need every workspace to depend on `react` and `@mdx-js/react` which is not very reasonable. For now I use `pnpFallbackMode: all` so they fallback to the `react` and `@mdx-js/react` in the top-level but again seems brittle.~~
- ~~Only resolve one entrypoint for each workspace. Tried multiple entries and some pages broke. Maybe try again later?~~
- ~~Some stuff have duplicate source links due to virtual paths. Need to deduplicate~~
- ~~Functions receive dedicated pages but the header is just "Callable". Ideally I want them to be just in the index page like variables~~

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by clemyan** [comment]
So I have resolved most of the issues, mostly by patching packages.

Only remaining issue is how to deal with workspaces that have multiple TS entrypoints. For now that's just only:

- `@yarnpkg/cli`
- `@yarnpkg/libzip`
- `@yarnpkg/pnp`
- `@yarnpkg/pnpify`

As far as I have tested, the default support for multiple entries in `docusaurus-plugin-typedoc-api` has a number of limitations:

- One must declare one of the entrypoints as the "index", which is what is shown as the package's Overview and its exports are shown at the top-level of the package in the sidebar
- Other entrypoints do not get their own Overview page and must be second-level in the sidebar, though its label can be change freely

image

I think I will try to further patch `docusaurus-plugin-typedoc-api` to override those restrictions.

**Comment by clemyan** [comment]
I may have gone overzealous with my patches 😅

image

Additions to the `docusaurus-plugin-typedoc-api` patch I made since my last comment:

- Each entrypoint gets its own Overview page, instead of just the main entrypoint
- Instead of pushing the non-main entrypoints to one level deeper in the sidebar than the main entrypoint, multiple entrypoints in the same package get their own dropdown, including the main entrypoint
- Add a level of non-collapsible headers to packages, so that we can make it just like the v3 API reference
- Add support for remark/rehype plugins, so that we can inject our AutoLink and CommandLineHighlight plugins

At this point I think the PR is ready

File: .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
++  packageCategories: [],
+   projectRoot: '.',
+   sortPackages: (a, d) => a.packageName.localeCompare(d.packageName),
+   sortSidebar: (a, d) => a.localeCompare(d),
+@@ -45,6 +46,8 @@ const DEFAULT_OPTIONS = {
+   routeBasePath: 'api',
+   tsconfigName: 'tsconfig.json',
+   typedocOptions: {},
++  remarkPlugins: [],
++  rehypePlugins: [],
```
Comment (arcanis) [inline .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch:27]: By any chance, can you try sending a PR upstream to get these changes accepted? I'd prefer if we could avoid maintaining permanent patches.

File: .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ const logger_1 = tslib_1.__importDefault(require("@docusaurus/logger"));
+ const utils_1 = require("@docusaurus/utils");
+@@ -200,8 +201,8 @@ ${assets ? `export const assets = ${createAssetsExportCode(assets)};` : ''}
+ `;
+     const code = `
+ ${pragma}
+-import React from 'react';
+-import { mdx } from '@mdx-js/react';
++import React from "${(0, url_1.pathToFileURL)(require.resolve('react'))}";
++import { mdx } from "${(0, url_1.pathToFileURL)(require.resolve('@mdx-js/react'))}";
```
Comment (merceyz) [inline .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch:29]: Would this allow removing the dependencies added in https://github.com/yarnpkg/berry/pull/5625/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519?

If you could send a PR upstream to fix this as well that would be great.

File: .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch (hunk 1 above)
Comment (clemyan) [reply .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch:27]: Done. milesj/docusaurus-plugin-typedoc-api#123

File: .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch (hunk 2 above)
Comment (clemyan) [reply .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch:29]: > Would this allow removing the dependencies added in https://github.com/yarnpkg/berry/pull/5625/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519?

Hmm.. Probably. I will test later

> If you could send a PR upstream to fix this as well that would be great.

I will submit one. Though I do think this code pattern is kinda obscure and won't be surprised if it is rejected.

File: .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch (hunk 1 above)
Comment (arcanis) [reply .yarn/patches/docusaurus-plugin-typedoc-api-npm-3.0.1-7dd061feb2.patch:27]: Thanks!

File: .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch (hunk 2 above)
Comment (merceyz) [reply .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch:29]: @josh-cena Indicated that they would accept such a change in https://discord.com/channels/226791405589233664/654372321225605128/949824347529445487

Comment (Josh-Cena) [reply .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch:29]: By the way, Docusaurus now uses MDX v3 and React automatic runtime, so things may be different ;)

Comment (clemyan) [reply .yarn/patches/@docusaurus-mdx-loader-npm-2.4.3-0d95b4751a.patch:29]: Yeah. Last night I started looking into making a PR and found that part of the MDX loader in Docusaurus v3 has changed significantly.

@merceyz Let's wait until after we migrate to Docusaurus v3 to see whether we still need the patch and/or to make an upstream PR

**Pull request merged**